main.go

```diff
@@ -951,3 +903,252 @@ func addControllers(
 		os.Exit(1)
 	}
 }
+
+// gatekeeperInstallationChecker starts or stops the gatekeeper-constraint-status-sync controller if Gatekeeper's
+// installation status changes. The controller must be off when Gatekeeper is not installed.
+func gatekeeperInstallationChecker(
+	ctx context.Context, managedCfg *rest.Config, dynamicClient dynamic.Interface, mgrOptions manager.Options,
+) error {
+	fieldSelector := fmt.Sprintf("metadata.name=constrainttemplates.%s", utils.GvkConstraintTemplate.Group)
+	timeout := int64(30)
+	crdGVR := schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}
+
+	newRetryWatcher := func() (*watch.RetryWatcher, bool, error) {
+		listResult, err := dynamicClient.Resource(crdGVR).List(
+			ctx, metav1.ListOptions{FieldSelector: fieldSelector, TimeoutSeconds: &timeout},
+		)
+		if err != nil {
+			return nil, false, err
+		}
+
+		gatekeeperInstalled := len(listResult.Items) > 0
+		resourceVersion := listResult.GetResourceVersion()
+
+		watchFunc := func(options metav1.ListOptions) (apiWatch.Interface, error) {
+			options.FieldSelector = fieldSelector
+
+			return dynamicClient.Resource(crdGVR).Watch(ctx, options)
+		}
+
+		w, err := watch.NewRetryWatcher(resourceVersion, &apiCache.ListWatch{WatchFunc: watchFunc})
+
+		return w, gatekeeperInstalled, err
+	}
+
+	healthAddress, err := getFreeLocalAddr()
+	if err != nil {
+		return err
+	}
+
+	// Disable the metrics endpoint for this manager. Note that since they both use the global
+	// metrics registry, metrics for this manager are still exposed by the other manager.
+	mgrOptions.Metrics.BindAddress = "0"
+	mgrOptions.LeaderElectionID = "governance-policy-framework-addon3.open-cluster-management.io"
+	mgrOptions.HealthProbeBindAddress = healthAddress
+
+	startMgr := func(ctx context.Context) {
```

Comment:
I think this could be defined outside of the `gatekeeperInstallationChecker` function. It might need more parameters, but I think it would help readability a lot.